pkg/db/db.go

```diff
@@ -0,0 +1,107 @@
+package db
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/red-hat-storage/managed-fusion-fleet-reconciler/pkg/types"
+	"go.uber.org/zap"
+)
+
+type DB struct {
+	Host      string `yaml:"host"`
+	Port      int    `yaml:"port"`
+	User      string `yaml:"user"`
+	Password  string `yaml:"password"`
+	Name      string `yaml:"name"`
+	TableName string `yaml:"providers_table_name"`
+}
+
+func getConnectionString(db *DB) string {
+	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", db.Host, db.Port, db.User, db.Password, db.Name)
+}
+
+type Postgres struct {
+	db   *DB
+	pool *pgxpool.Pool
+}
+
+func NewPGInstance(ctx context.Context, logger *zap.Logger, db *DB) (*Postgres, error) {
+	pool, err := pgxpool.New(ctx, getConnectionString(db))
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	return &Postgres{db: db, pool: pool}, nil
+}
+
+func (pg *Postgres) Ping(ctx context.Context) error {
+	return pg.pool.Ping(ctx)
+}
+
+func (pg *Postgres) Close() {
+	pg.pool.Close()
+}
+
+func (pg *Postgres) SetupNotification(ctx context.Context, reqChan chan types.Request) (*pgconn.PgConn, error) {
+	config, err := pgconn.ParseConfig(getConnectionString(pg.db))
+	if err != nil {
+		return nil, err
+	}
+	config.OnNotification = func(c *pgconn.PgConn, n *pgconn.Notification) {
+		req := types.Request{}
+		if err := json.Unmarshal([]byte(n.Payload), &req); err != nil {
+			return
+		}
+		reqChan <- req
+	}
+
+	pgConn, err := pgconn.ConnectConfig(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+	query := fmt.Sprintf("LISTEN %s;", pg.db.TableName)
+	// Listen for notifications
+	_, err = pgConn.Exec(ctx, query).ReadAll()
+	if err != nil {
+		pgConn.Close(ctx)
+	}
+
+	return pgConn, err
+}
+
+func (pg *Postgres) GetProviderCluster(ctx context.Context, clusterID string) (*types.ProviderCluster, error) {
+	query := fmt.Sprintf("SELECT * FROM %s WHERE cluster_id = $1", pg.db.TableName)
+	row := pg.pool.QueryRow(ctx, query, clusterID)
+
+	var c types.ProviderCluster
+	err := row.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status)
+	if err != nil {
+		return nil, err
+	}
+	return &c, nil
+}
+
+func (pg *Postgres) ListProviderClusters(ctx context.Context) ([]*types.ProviderCluster, error) {
+	query := fmt.Sprintf("SELECT * FROM %s", pg.db.TableName)
+	rows, err := pg.pool.Query(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var cluster []*types.ProviderCluster
+	for rows.Next() {
+		var c types.ProviderCluster
+		err := rows.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status)
+		if err != nil {
+			return nil, err
+		}
+		cluster = append(cluster, &c)
+	}
```

Comment:
As discussed in the last review, we do not want to load all providers into memory at the same time